### PR TITLE
StarRocks support kerberos authentication

### LINKED ISSUE #2602: StarRocks support kerberos authentication
## Feature request

**Is your feature request related to a problem? Please describe.**

Kerberos is a computer network authentication protocol that allows an entity to communicate in a non-secure network environment and prove its identity to another entity in a secure way. Since Hadoop-related applications all support the kerberos protocol, StarRocks Enterprise Edition also needs to provide support for the kerberos protocol.

**Describe the solution you'd like**

Only mysql client 8.0.26 and above versions support this function.
Support dynamic configuration, no need to restart fe.

You would add follow configuration by "admin set frontend config" clause or fe.conf:
authentication_kerberos_enabled=true
authentication_kerberos_service_principal=<YOUR_SERVICE_PRINCIPAL>
authentication_kerberos_service_key_tab=<YOUR_SERVICE_PRINCIPAL_KEYTAB_FILE_PATH>

Syntax:
```sql
CREATE USER
user_identity [auth_option]
[DEFAULT ROLE 'role_name']

user_identity:
    'user_name'@'host'

auth_option: {
    IDENTIFIED BY 'auth_string'
    IDENTIFIED WITH auth_plugin
    IDENTIFIED WITH auth_plugin BY 'auth_string'
    IDENTIFIED WITH auth_plugin AS 'auth_string'
}
```

for example:
```sql
CREATE USER 'stephen'
  IDENTIFIED WITH authentication_kerberos
```
or
```sql
CREATE USER 'stephen'
  IDENTIFIED WITH authentication_kerberos
  BY 'STARROCKS.COM'
```

mysql client 8.0.26 connect to starrocks fe:
If you have cached TGT, you can execute follow command:
```
mysql --default-auth=authentication_kerberos_client -u stephen
```
or:
```
mysql --default-auth=authentication_kerberos_client -u stephen -p
```
then input your upn password.
The above principals need to be registered in advance with kdc

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by stephen-shelby**
Kerberos is a computer network authentication protocol that allows an entity to communicate in a non-secure network environment and prove its identity to another entity in a secure way. Since Hadoop-related applications all support the Kerberos protocol, StarRocks Enterprise Edition also needs support for the Kerberos protocol.

usage:
- Only MySQL client 8.0.26 and above versions support Kerberos authentication.
- Only enterprise edition support Kerberos authentication.
- Support dynamic configuration and classes; no need to restart FE.
- You would add the following configuration by "admin set frontend config" clause or fe.conf:
   1. enable_authentication_kerberos=true
   2. authentication_kerberos_service_principal=<YOUR_SERVICE_PRINCIPAL>
   3. authentication_kerberos_service_key_tab=<YOUR_SERVICE_PRINCIPAL_KEYTAB_FILE_PATH>

syntax:
```SQL
CREATE USER
user_identity [auth_option]
[DEFAULT ROLE 'role_name']

user_identity:
    'user_name'@'host'

auth_option: {
    IDENTIFIED BY 'auth_string'
    IDENTIFIED WITH auth_plugin
    IDENTIFIED WITH auth_plugin BY 'auth_string'
    IDENTIFIED WITH auth_plugin AS 'auth_string'
}
```
The syntax remains the same, but `authentication_kerberos` is added to auth_plugin.
If you do not specify user realm to create a user, use service principal realm as the default value.

connect to starrocks fe:
If you have cached TGT, you can execute the following command:

```
mysql --default-auth=authentication_kerberos_client -u stephen
```
or
```
mysql --default-auth=authentication_kerberos_client -u stephen -p
```
Then input your user principal password.

The above principles need to be registered in advance with the KDC server.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateUserStmt.java [hunk 1]
```diff
@@ -162,6 +164,12 @@ public void analyze(Analyzer analyzer) throws UserException {
                 } else {
                     scramblePassword = new byte[0];
                 }
+            } else if (AuthPlugin.AUTHENTICATION_KERBEROS.name().equals(authPlugin) && Auth.isSupportKerberosAuth()) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateUserStmt.java:167]: equals or ignorecaseEquals?

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateUserStmt.java [hunk 2]
```diff
@@ -162,6 +164,12 @@ public void analyze(Analyzer analyzer) throws UserException {
...
+                if (userForAuthPlugin != null) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateUserStmt.java:168]: Could you provide some explanation for this judge?

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 3]
```diff
@@ -1153,6 +1153,24 @@
     @ConfField
     public static String auth_token = "";
 
+    /**
+     * If set to true and the jar that use to authentication is loaded in fe, kerberos authentication is supported.
+     */
+    @ConfField(mutable = true)
+    public static boolean authentication_kerberos_enabled = false;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:1160]: How about enable_authentication_kerberos?
All switch configuration stars with `enable_`

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 4]
```diff
@@ -1153,6 +1153,24 @@
...
+
+    /**
+     * Service principal name. like "starrocks-fe/hostname@STARROCKS.COM"
+     */
+    @ConfField(mutable = true)
+    public static String authentication_kerberos_service_principal = "";
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:1166]: Why is this config needed? Should add more detail explanation.

File: fe/fe-core/src/main/java/com/starrocks/common/FeNameFormat.java [hunk 5]
```diff
@@ -33,6 +33,7 @@
     // so it can not distinguish whether it is an operator or a column name
     // the future new design will improve this problem and open this limitation
     private static final String COLUMN_NAME_REGEX = "^[^\0=<>!\\*]{1,64}$";
+    private static final String MYSQL_USER_NAME_REGEX = "^[a-zA-Z][a-zA-Z0-9_]{1,63}/?[.a-zA-Z0-9_-]{0,63}$";
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/common/FeNameFormat.java:36]: Is this the same as MySQL's name regex? Better to add more comment

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/UserPrivTable.java [hunk 6]
```diff
@@ -71,6 +71,26 @@ public Password getPassword(UserIdentity currentUser) {
         return null;
     }
 
+    public Password getPasswordByApproximate(String remoteUser, String remoteHost) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/UserPrivTable.java:74]: Give some comment for this function

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 7]
```diff
@@ -1396,6 +1400,16 @@ public static Auth read(DataInput in) throws IOException {
...
     }
 
+    public static boolean isSupportKerberosAuth() {
+        if (!Config.authentication_kerberos_enabled ||
+                Config.authentication_kerberos_service_key_tab.isEmpty() ||
+                Config.authentication_kerberos_service_principal.isEmpty()) {
+            return false;
+        }
+
+        return new File(KRB5_AUTH_JAR_PATH).exists();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:1410]: This is a heavy operation.
is it necessary to be checked?

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 8]
```diff
@@ -73,6 +75,9 @@
     public static final String ROOT_USER = "root";
     public static final String ADMIN_USER = "admin";
 
+    public static final String KRB5_AUTH_CLASS_NAME = "com.starrocks.auth.KerberosAuthentication";
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:78]: How about "com.starrocks.plugins.auth.KerberosAuthentication"?

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java (hunk 3 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java]: done

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java (hunk 4 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:1171]: done

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java (hunk 7 above)
Comment (imay) [reply fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:1410]: IMO the JAR file need to be loaded only once. Could you optimize it?

File: fe/fe-core/src/main/java/com/starrocks/mysql/MysqlHandshakePacket.java [hunk 9]
```diff
@@ -104,4 +122,28 @@ public void buildAuthSwitchRequest(MysqlSerializer serializer) {
...
+            LOG.error(msg);
+            throw new Exception(msg);
+        }
+
+        String userRealm = password.getUserForAuthPlugin();
+        File jarFile = new File(KRB5_AUTH_JAR_PATH);
+        ClassLoader loader = URLClassLoader.newInstance(
+                new URL[] { jarFile.toURL() },
+                getClass().getClassLoader()
+        );
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/mysql/MysqlHandshakePacket.java]: Better to load the class only once. One option is to create an interface in fe, and the implementation of the interface is in the commercial version of the jar package.

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java (hunk 7 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java]: done

File: fe/fe-core/src/main/java/com/starrocks/mysql/MysqlHandshakePacket.java (hunk 9 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/mysql/MysqlHandshakePacket.java]: done

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 10]
```diff
@@ -1390,6 +1397,39 @@ public TFetchResourceResult toResourceThrift() {
...
+                LOG.error("Failed to load {}", Auth.KRB5_AUTH_CLASS_NAME, e);
+                return false;
+            }
+        }
+    }
+
+    public Class<?> getAuthClazz() {
+        return authClazz;
+    }
+
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:1432]: This is not thread-safe, right?

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 11]
```diff
@@ -1390,6 +1397,39 @@ public TFetchResourceResult toResourceThrift() {
         }
     }
 
+    public  boolean isSupportKerberosAuth() {
+        if (!Config.enable_authentication_kerberos ||
+                Config.authentication_kerberos_service_key_tab.isEmpty() ||
+                Config.authentication_kerberos_service_principal.isEmpty()) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:1403]: Add more concise error messages to dedicate the error.

**Comment by chaoyli** [comment]
Commit message to add more question, solution, and usage.

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java (hunk 11 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java]: done

**Comment by stephen-shelby** [comment]
> Commit message to add more question, solution, and usage.

done

**Review by gengjun-git — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**